changefeedccl: reduce use of errors during pauseOrResumePolling

wenyihu6 opened the pull request:
**schemafeed: add BenchmarkPauseOrResumePolling**

This patch introduces a benchmark for pauseOrResumePolling, as it’s an important
function in a hot path that has previously caused performance regressions due to
complex error handling.

Epic: none
Release note: none

---

**changefeedccl: reduce use of errors during pauseOrResumePolling**

Previously, errors were constructed and compared in pauseOrResumePolling solely
to indicate that pausing cannot be stopped and period table history scan is
necessary during pauseOrResumePolling. However, no actual errors were returned
from pauseOrResumePolling. This error handling introduced unnecessary CPU
overhead in this hot path, as observed during escalations and DRT scale testing.
This patch improves performance by replacing error-based communication with a
boolean check.

Fixes: https://github.com/cockroachdb/cockroach/issues/131327
Release note (performance improvement): Enhanced performance when schema_locked
is not in use by improving error handling during periodic table history polling.

```
                                        │  before.txt  │              after.txt               │
                                        │    sec/op    │    sec/op     vs base                │
PauseOrResumePolling/non-terminal_error    5.533µ ± 5%   3.849µ ± 29%  -30.44% (p=0.000 n=10)
PauseOrResumePolling/not_schema_locked    7069.5n ± 4%   176.4n ±  9%  -97.50% (p=0.000 n=10)
PauseOrResumePolling/schema_locked         136.2n ± 4%   131.8n ±  7%        ~ (p=0.306 n=10)
```

wenyihu6 committed:
changefeedccl: reduce use of errors during pauseOrResumePolling

Previously, errors were constructed and compared in pauseOrResumePolling solely
to indicate that pausing cannot be stopped and period table history scan is
necessary during pauseOrResumePolling. However, no actual errors were returned
from pauseOrResumePolling. This error handling introduced unnecessary CPU
overhead in this hot path, as observed during escalations and DRT scale testing.
This patch improves performance by replacing error-based communication with a
boolean check.

Fixes: #131327
Release note (performance improvement): Enhanced performance when schema_locked
is not in use by improving error handling during periodic table history polling.

```
❯ benchstat before.txt after.txt
goos: darwin
goarch: arm64
                                        │  before.txt  │              after.txt               │
                                        │    sec/op    │    sec/op     vs base                │
PauseOrResumePolling/non-terminal_error    5.533µ ± 5%   3.849µ ± 29%  -30.44% (p=0.000 n=10)
PauseOrResumePolling/not_schema_locked    7069.5n ± 4%   176.4n ±  9%  -97.50% (p=0.000 n=10)
PauseOrResumePolling/schema_locked         136.2n ± 4%   131.8n ±  7%        ~ (p=0.306 n=10)
geomean                                    1.747µ        447.2n        -74.39%
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131951)
<!-- Reviewable:end -->

stevendanna commented on pkg/ccl/changefeedccl/schemafeed/schema_feed.go:
```diff
@@ -480,36 +480,36 @@ func (tf *schemaFeed) pauseOrResumePolling(ctx context.Context, atOrBefore hlc.T
 
-		return nil
-	}); err != nil {
+	if !canPausePolling {
```
In case this changes in the future, should we consider `if (!canPausePolling || err != nil)` ?

wenyihu6 replied:
Good idea, done.

stevendanna commented on pkg/ccl/changefeedccl/schemafeed/schema_feed.go:
```diff
@@ -480,36 +480,36 @@ func (tf *schemaFeed) pauseOrResumePolling(ctx context.Context, atOrBefore hlc.T
 		if desc1.GetVersion() != desc2.GetVersion() {
-			return errors.Newf("desc %d version changed from version %d to %d between frontier %s and atOrBefore %s",
-				desc1.GetID(), desc1.GetVersion(), desc2.GetVersion(), frontier, atOrBefore)
+			return false, nil
```
If we didn't want to lose this log message, I suppose we could put it behind `if log.V(2) { ... }`. I'm not sure what that'll do to your benchmark since we've see `log.V()` show up in profiles before.  I'm not sure howe often we've used this log message for debugging.

wenyihu6 replied:
Good idea, done.

stevendanna reviewed (commented):
Nice!

stevendanna commented on pkg/ccl/changefeedccl/schemafeed/schema_feed.go:
```diff
@@ -480,36 +480,37 @@ func (tf *schemaFeed) pauseOrResumePolling(ctx context.Context, atOrBefore hlc.T
 		desc1 := ld1.Underlying().(catalog.TableDescriptor)
 		if !desc1.IsSchemaLocked() {
-			return errors.Newf("desc %d not schema-locked at frontier %s", desc1.GetID(), frontier)
+			log.VInfof(ctx, 2, "desc %d not schema-locked at frontier %s", desc1.GetID(), frontier)
```
Would have to look at the benchmark to be sure, but my guess is that putting this behind `if log.V(2)` would help in the common case. But, it probably doesn't matter too much as you've already improved this function substantially.

stevendanna replied:
I know I am the one who recommended this, but I wonder if there is really a world in production where we'd want to print this on every message. I know we were doing it before implicitly by logging the error at V(1) so I leave it to y'alls judgement.

wenyihu6 replied:
> my guess is that putting this behind if log.V(2)

Using V(1) sounds good to me which matches what we do for other error cases https://github.com/cockroachdb/cockroach/blob/f701074bdb42dbd7e078b3ce8a0ee4cbaca82fab/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L527-L529.

stevendanna reviewed (approved):
Nice improvement.

> , so I'm skipping tests for this PR

That seems reasonable to me. If you did want to add a benchmark, I _think_ you could probably use that test as the basis of a benchmark:

```golang
func BenchmarkPauseOrResumePolling(b *testing.B) {
	defer leaktest.AfterTest(b)()
	defer log.Scope(b).Close(b)

	ctx := context.Background()

	const tableID = 123
	sf := schemaFeed{
		leaseMgr: &testLeaseAcquirer{
			id: tableID,
			descs: []*testLeasedDescriptor{
				newTestLeasedDescriptor(tableID, 1, false, hlc.Timestamp{WallTime: 20}),
				newTestLeasedDescriptor(tableID, 2, true, hlc.Timestamp{WallTime: 40}),
			},
		},
		targets: CreateChangefeedTargets(tableID),
	}
	setFrontier := func(ts hlc.Timestamp) error {
		sf.mu.Lock()
		defer sf.mu.Unlock()
		return sf.mu.ts.advanceFrontier(ts)
	}

	// Set the initial frontier to 10.
	require.NoError(b, setFrontier(hlc.Timestamp{WallTime: 10}))
	require.False(b, sf.pollingPaused())

	b.Run("non-terminal error", func(b *testing.B) {
		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			require.NoError(b, sf.pauseOrResumePolling(ctx, hlc.Timestamp{WallTime: 10}))
		}
		require.False(b, sf.pollingPaused())
	})
	b.Run("not schema locked", func(b *testing.B) {
		require.NoError(b, setFrontier(hlc.Timestamp{WallTime: 30}))
		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			require.NoError(b, sf.pauseOrResumePolling(ctx, hlc.Timestamp{WallTime: 30}))
		}
		require.False(b, sf.pollingPaused())
	})
	b.Run("schema locked", func(b *testing.B) {
		require.NoError(b, setFrontier(hlc.Timestamp{WallTime: 50}))
		for i := 0; i < b.N; i++ {
			require.NoError(b, sf.pauseOrResumePolling(ctx, hlc.Timestamp{WallTime: 50}))
		}
		require.True(b, sf.pollingPaused())
	})
}
```

At least, it shows the improvement here nicely:

```
                                        │  before.txt  │              after.txt              │
                                        │    sec/op    │   sec/op     vs base                │
PauseOrResumePolling/non-terminal_error    5.474µ ± 1%   3.726µ ± 1%  -31.93% (p=0.000 n=10)
PauseOrResumePolling/not_schema_locked    6977.0n ± 4%   211.0n ± 0%  -96.98% (p=0.000 n=10)
PauseOrResumePolling/schema_locked         140.3n ± 2%   138.9n ± 2%        ~ (p=0.210 n=10)
geomean                                    1.750µ        478.0n       -72.68%

                                        │   before.txt   │               after.txt                │
                                        │      B/op      │     B/op      vs base                  │
PauseOrResumePolling/non-terminal_error   1.875Ki ± 0%     1.141Ki ± 0%  -39.17% (p=0.000 n=10)
PauseOrResumePolling/not_schema_locked    2216.00 ± 0%       16.00 ± 0%  -99.28% (p=0.000 n=10)
PauseOrResumePolling/schema_locked          0.000 ± 0%       0.000 ± 0%        ~ (p=1.000 n=10) ¹
geomean                                                ²                 -83.62%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                                        │  before.txt   │              after.txt               │
                                        │   allocs/op   │ allocs/op   vs base                  │
PauseOrResumePolling/non-terminal_error    43.00 ± 0%     24.00 ± 0%  -44.19% (p=0.000 n=10)
PauseOrResumePolling/not_schema_locked    45.000 ± 0%     1.000 ± 0%  -97.78% (p=0.000 n=10)
PauseOrResumePolling/schema_locked         0.000 ± 0%     0.000 ± 0%        ~ (p=1.000 n=10) ¹
geomean                                               ²               -76.85%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean
```

wenyihu6 committed:
schemafeed: add BenchmarkPauseOrResumePolling

This patch introduces a benchmark for pauseOrResumePolling, as it’s an important
function in a hot path that has previously caused performance regressions due to
complex error handling.

Epic: none
Release note: none

wenyihu6 commented:
Thanks for the thorough review!

bors r=stevendanna
